tmck-code/csgo-surf-local-server
# Fix dockerfile

There were quite a few typos and mismatches in the various files which I've tidied up here, namely

- deciding on `csgo-data` as the bind-mounted host dir, which mounts to `/home/steam/csgo-dedicated`
- fixing up a typo in the paths for the DB init scripts
- patching the `entry.sh` from `https://github.com/CM2Walki/CSGO/blob/master/bullseye/etc/entry.sh` to resolve a very frustrating error

## The entry.sh bug

After the release of CS2, the CSGO server seems to have a bug where
- the server thinks it needs an update
- restarts
- doesn't update
- go to step 1 and repeat

I tried looking at the CSGO/sourcemod/individual plugin versions, but the fix turned out to be ignoring the message entirely :facepalm:.

This is done by replacing the `-autoupdate` arg with `-norestart`, and now the server works fine! :tada: 

## TODO:

- [x] make a healthcheck for the mysql container so that the `db/bootstrap` task doesn't fail
- [x] sort out the `plugin/bootstrap`
- [x] move to this dockerfile? as it doesn't contain the entry.sh bug AFAICS https://github.com/timche/docker-csgo

---

## Update 2025

Hello, me from the past! I've attempted to pick this up again and have hit quite a few roadbumps as it's been over a year since I did anything with this.

As it stands, I _think_ everything should work, maybe, but I need to properly document everything that's going on as there are a lot of moving pieces now.

Currently, all I've managed to do is to generate a fresh GSLT token, and run a CSGO server with the `cm2network/csgo:latest` (non-sourcemod) image. IIRC sourcemod was giving a lot of trouble.

1. Run the docker container with
    ```shell
    docker run -it --rm \
        --net=host \
        -v $(pwd)/csgo-data:/home/steam/csgo-dedicated \
        --name surftimer \
        -e SRCDS_TOKEN=$CSGO_GSLT \
        -e ADDITIONAL_ARGS="-norestart" \
        cm2network/csgo:latest bash
    ```
2. Run the CSGO server inside the container with
    ```shell
    ./csgo-dedicated/srcds_run \
    -game csgo \
    -console \
    -usercon +game_type 0 +game_mode 0 +mapgroup mg_active +map de_dust2 \
    +sv_setsteamaccount $SRCDS_TOKEN
    ```

- [x] I can successfully SEE the server from the in-game server browser via `10.0.LOCAL_IP:27015`,
- [x] I can get a response from the rcon server :raised_hands: with `RCON_HOST=THE_SAME_IP RCON_PASSWORD=guessed ./ops/scripts/rcon_command version`
  - `rcon.exceptions.WrongPassword`
  - this is much more than the `ConnectionRefusedError: [Errno 111] Connection refused` that I usually get


Updates to come! :crossed_fingers: 

PS. I think the move to `timche/docker-csgo` image was possibly a mistake given that everything seems to be borked now, but I'll see :thinking: 

## Update 2025-06-01

In the `server.cfg` this arg must be set if using a local IP!
- `sv_lan "1"`

## Update 2025-09-18

When bootstrapping a server for the first time, a few things need to happen in sequence

1. Run the `cm2network/csgo:latest` container, bind-mounting it to `csgo-dedicated` in the host dir. (the gslt token isn't strictly required for this stage)
  ```shell
  docker run -it \
    --rm \
    --net=host \
    -v $(pwd)/csgo-dedicated:/home/steam/csgo-dedicated \
    --name surftimer \
    -e SRCDS_TOKEN=$CSGO_GSLT \
    -e ADDITIONAL_ARGS="-norestart" \
    cm2network/csgo:latest \
    bash
  ```
2. Inside the container, take advantage of the preprepared ENV vars, and run this command which will download all the server files, including the `srcds_run` binary which will become the usual way to run the server
  ```shell
  bash "${STEAMCMDDIR}/steamcmd.sh" \
    +force_install_dir "${STEAMAPPDIR}" \
    +login anonymous \
    +app_update "${STEAMAPPID}" validate \
    +quit
  ```
3. Create a file `csgo-dedicated/csgo/cfg/server.cfg`
  ```
  sv_lan 1
  rcon_password "password_here"
  ```
  - `sv_lan 1` when accessing the server, it will immediately exit to the main menu unless this is set
  - set the rcon password to avoid confusion
4. Now you can run the server with the command from above^^
  ```shell
  ./csgo-dedicated/srcds_run \
    -game csgo \
    -console \
    -usercon +game_type 0 +game_mode 0 +mapgroup mg_active +map de_dust2 \
    +sv_setsteamaccount $SRCDS_TOKEN
  ```
5. Check out some rcon commands
  ```shell
  # get server status
  RCON_HOST=10.0.0.34 RCON_PASSWORD=yolo ./ops/scripts/rcon_command status
  # kick all bots
  RCON_HOST=10.0.0.34 ./ops/scripts/rcon_command 'bot_kick'
  # shutdown the server
  RCON_HOST=10.0.0.34 ./ops/scripts/rcon_command _restart
  # list all maps
  RCON_HOST=10.0.0.34 ./ops/scripts/rcon_command 'maps *'
  ```

## Update 2025-09-20

- [As per the README](https://github.com/surftimer/SurfTimer?tab=readme-ov-file#installation), SurfTimer requires the mysql db to us the `"mysql_native_password authentication/method"`.
  - This is done by passing the flag as the "command" in the docker compose definition (a little strange), e.g.
  - `command: --default-authentication-plugin=mysql_native_password`
- [According to the Mysql documentation](https://dev.mysql.com/doc/refman/8.4/en/native-pluggable-authentication.html), this auth configuration has been completely removed in the current versions (8.4+)
  - It is marked as deprecated in 8.0.43, so I've had to peg mysql to that version in order for this to work
- The same doc also details how the auth works nowadays, but it seems like a different client-side arg is needed to use it, which would require changing SurfTimer code, and I'm not quite there yet :grimacing:
